cumulus/primitives/storage-weight-reclaim/src/lib.rs

```diff
@@ -0,0 +1,516 @@
+		let current_storage_proof_size = get_proof_size()?;
+		let previous_storage_proof_size = self.previous_reported_proof_size?;
+		let used_weight =
+			self.previous_remaining_proof_size.saturating_sub(current_remaining_weight);
+		let reported_used_size =
+			current_storage_proof_size.saturating_sub(previous_storage_proof_size);
+		let reclaimable = used_weight.saturating_sub(reported_used_size);
+		log::trace!(
+			target: LOG_TARGET,
+			"Found reclaimable storage weight. benchmarked: {used_weight}, consumed: {reported_used_size}"
+		);
+
+		self.previous_remaining_proof_size = current_remaining_weight.saturating_add(reclaimable);
+		self.previous_reported_proof_size = Some(current_storage_proof_size);
+		Some(Weight::from_parts(0, reclaimable))
+	}
+
+	/// Check the consumed storage weight and add the reclaimed
+	/// weight budget back to `weight_meter`.
+	pub fn reclaim_with_meter(&mut self, weight_meter: &mut WeightMeter) -> Option<Weight> {
+		let reclaimed = self.reclaim(weight_meter.remaining())?;
+		weight_meter.reclaim_proof_size(reclaimed.proof_size());
+		Some(reclaimed)
+	}
+}
+
+/// Returns the current storage proof size from the host side.
+///
+/// Returns `None` if proof recording is disabled on the host.
+pub fn get_proof_size() -> Option<u64> {
+	let proof_size = storage_proof_size();
+	(proof_size != PROOF_RECORDING_DISABLED).then_some(proof_size)
+}
+
+/// Storage weight reclaim mechanism.
+///
+/// This extension checks the size of the node-side storage proof
+/// before and after executing a given extrinsic. The difference between
+/// benchmarked and spent weight can be reclaimed.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, Default, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct StorageWeightReclaim<T: Config + Send + Sync>(PhantomData<T>);
+
+impl<T: Config + Send + Sync> core::fmt::Debug for StorageWeightReclaim<T> {
+	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+		let _ = write!(f, "StorageWeightReclaim");
+		Ok(())
+	}
+}
+
+impl<T: Config + Send + Sync> SignedExtension for StorageWeightReclaim<T>
+where
+	T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
+{
+	const IDENTIFIER: &'static str = "StorageWeightReclaim";
+
+	type AccountId = T::AccountId;
+	type Call = T::RuntimeCall;
+	type AdditionalSigned = ();
+	type Pre = Option<u64>;
+
+	fn additional_signed(
+		&self,
+	) -> Result<Self::AdditionalSigned, sp_runtime::transaction_validity::TransactionValidityError>
+	{
+		Ok(())
+	}
+
+	fn pre_dispatch(
+		self,
+		_who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &sp_runtime::traits::DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> Result<Self::Pre, sp_runtime::transaction_validity::TransactionValidityError> {
+		Ok(get_proof_size())
+	}
+
+	fn post_dispatch(
+		pre: Option<Self::Pre>,
+		info: &DispatchInfoOf<Self::Call>,
+		_post_info: &PostDispatchInfoOf<Self::Call>,
```

Comment:
If the call returned some weight, [this logic](https://github.com/paritytech/polkadot-sdk/blob/e675bb44d5d9840f11dfa6f23bfeabc27589cbf5/substrate/frame/system/src/extensions/check_weight.rs#L262-L266) will be triggered in `CheckWeight`. This would ultimately lead to reducing the proof size twice. Thus, we need to take this into account here.

Reply:
Otherwise we would require to have a forced order of `CheckWeight` and this extension.